### DISCUSSION TITLE
FlowMap Parsing is quite different between v1.10.2 and v2.0.0-5

### DISCUSSION THREAD

**Discussion by ipapast** [eemeli/yaml#280] [Q&A] [answered]
Hi, 
I wanted to ask what changed betwen these versions and they now parse documents quite differently. 

My example yaml:

```
AWSTemplateFormatVersion: "2010-09-09"
Description: Template generated
Resources:
  Denied:
    Type: "AWS::IAM::Role"
    Properties:
      AssumeRolePolicyDocument: {
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "AWS": "arn:aws:iam::123456789012:root"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
```

Parsing with v1.10.2:

```
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "Template generated by rain",
  "Resources": {
    "Denied": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "Roles": [
          "test"
        ],
        "PolicyName": "test",
        "PolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Action": [
                "*"
              ],
              "Effect": "Allow",
              "Resource": "*"
            }
          ]
        }
      }
    }
  }
}

```
And the errors are:
-> doc.errors[0].message Insufficient indentation in flow collection
-> doc.errors[0].name YAMLSemanticError

In version 2.0.0-5, the parsed Document is:
```
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "Template generated by rain",
  "Resources": {
    "Denied": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "Roles": [
          "test"
        ],
        "PolicyName": "test",
        "PolicyDocument": {}
      }
    },
    "Version": {
      "2012-10-17": [
        {
          "Action": [
            "*"
          ],
          "Effect": "Allow",
          "Resource": "*"
        }
      ]
    }
  }
}
```
and the errors are:
-> doc.errors[0].message Expected flow map to end with } at line 10, column 24:
      PolicyDocument: {
                       ^
-> doc.errors[0].name YAMLParseError
-> doc.errors[0].code MISSING_CHAR

There are clearly differences in the way these are parsed, you can see how PolicyDocument is an {} in the second case.
I am wondering what is right/ wrong (I have a feeling the v1.10.2 parses it correctly), and if it's not a single answer what decisions/changes drove this. 

Thank you!

**Comment by eemeli** [comment, marked as answer]
When you embed a flow collection within a block collection, you still need to follow the indentation rules of the outer collection. In other words, all of the flow collection needs to be more indented than its parent. The final closing brace can be at the same level as the key, but other parsers might complain about that.

As for what changed, well, the parser was entirely rewritten from the ground up to include a lexer as the first step, and the intermediate CST structure is now completely different. In both cases, the parser is complaining about the input being broken, but the error messages are a bit different.

**Reply by ipapast** [reply]
Thank you. I went through the release docs and it makes a bit more sense. This can be closed now :)
